libia2/include/permissive_mode.h

```diff
@@ -1,10 +1,27 @@
 #ifndef _GNU_SOURCE
 #define _GNU_SOURCE // for SIG* macro
 #endif
+#include <assert.h>
+#include <dlfcn.h>
 #include <pthread.h>
 #include <signal.h>
+#include <stdbool.h>
 #include <stddef.h>
+#include <stdint.h>
 #include <stdio.h>
+#include <string.h>
+
+#include "ia2.h"
+
+/*
+ * PKRU is defined to be bit 9 in the XSAVE state component bitmap by the Intel
+ * manual.
+ */
+#define XFEATURE_PKRU 9
+#define XFEATURE_MASK_PKRU		(1 << XFEATURE_PKRU)
+
+/* Offset of the XSAVE header in bytes from the XSAVE area base */
+#define XSAVE_HEADER_OFFSET 512
```

Comment:
Where does this constant come from? The Intel [Software Developers Manual](https://www.intel.com/content/dam/www/public/us/en/documents/manuals/64-ia-32-architectures-software-developer-vol-1-manual.pdf) says the offset used in has to be looked up via cpuid (§13.4.3):
> As noted in Section 13.2, CPUID.(EAX=0DH,ECX=9):EBX enumerates the offset (in bytes, from the base of the
XSAVE area) of the section of the extended region of the XSAVE area used for PKRU state (when the standard
format of the extended region is used).

The alternative to the standard format is the compacted format, which requires adding up sizes for the saved components, and which I'm guessing isn't enabled/relevant here.

Reply:
This is not the PKRU offset (which we look up from cpuid). This is the offset of the XSAVE header which is always 512 bytes in.

> The XSAVE header of an XSAVE area comprises the 64 bytes starting at offset 512 from the area’s base address:
 Bytes 7:0 of the XSAVE header is a state-component bitmap (see Section 13.1) called XSTATE_BV. It
identifies the state components in the XSAVE area.

The compacted format is never used for the signal handler context here, fortunately. We could and maybe should check the relevant bit of the header to ensure this but :man_shrugging:.

Reply:
Ah, I very much had the wrong context in mind here. This is fine then, though it would be nice to point to the docs where this comes from (Intel manual §13.4.2).